Bug: Eslint plugin failed downloads

### Which package(s) does this bug affect?

- [ ] Create Plugin
- [ ] Sign Plugin
- [ ] Plugin E2E
- [ ] Plugin Meta Extractor
- [x] Plugin ESLint Rules

### Package versions

```shell
0.5.0
```

### What happened?

Due to changes in the structure of the bundled grafana npm packages (ui, data, runtime) depending on the version specified it can result in download failures.

```
Please wait... downloading Grafana types information for version 11.6.0.
Failed to download @grafana/data: HTTP status 404

Oops! Something went wrong! :(

ESLint: 9.28.0

Error: Error while loading rule '@grafana/plugins/import-is-compatible': Failed to download types for @grafana/data@11.6.0: Command failed: node -e "const https = require('https'); https.get('https://cdn.jsdelivr.net/npm/@grafana/data@11.6.0/dist/index.d.ts', (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
              if (res.statusCode === 200) {
                require('fs').writeFileSync('/var/folders/7f/1_46h_kj2d15rltqk2jzh81c0000gn/T/gf-eslint-plugin-compatible-11.6.0/node_modules/@grafana/data/index.d.ts', data);
                process.exit(0);
              }
              console.error('Failed to download @grafana/data: HTTP status ' + res.statusCode);
              process.exit(1);
            });
          }).on('error', (e) => {
            console.error('Failed to download @grafana/data: ' + e.message);
            process.exit(1);
          })"
Failed to download @grafana/data: HTTP status 404
```

Additionally versions after 12.1.0 contain multiple typescript definition files resulting in false positives where only the index.d.ts is downloaded rather than all the d.ts files:

```
Please wait... downloading Grafana types information for version 12.1.0.

/Users/jackwestbrook/dev/grafana/scenes/packages/scenes/src/components/SceneApp/SceneApp.test.tsx
  2:10  warning  The member "locationService" is not available in all runtime environments that this plugin supports. Make sure to check if the member is undefined before accessing it, or it may cause runtime errors. "@grafana/runtime@12.1.0" does not export member "locationService"  @grafana/plugins/import-is-compatible

/Users/jackwestbrook/dev/grafana/scenes/packages/scenes/src/components/SceneApp/utils.ts
  3:10  warning  The member "locationSearchToObject" is not available in all runtime environments that this plugin supports. Make sure to check if the member is undefined before accessing it, or it may cause runtime errors. "@grafana/runtime@12.1.0" does not export member "locationSearchToObject"  @grafana/plugins/import-is-compatible

/Users/jackwestbrook/dev/grafana/scenes/packages/scenes/src/components/SceneRefreshPicker.test.ts
  6:10  warning  The member "config" is not available in all runtime environments that this plugin supports. Make sure to check if the member is undefined before accessing it, or it may cause runtime errors. "@grafana/runtime@12.1.0" does not export member "config"  @grafana/plugins/import-is-compatible

/Users/jackwestbrook/dev/grafana/scenes/packages/scenes/src/components/SceneRefreshPicker.tsx
  4:10  warning  The member "config" is not available in all runtime environments that this plugin supports. Make sure to check if the member is undefined before accessing it, or it may cause runtime errors. "@grafana/runtime@12.1.0" does not export member "config"  @grafana/plugins/import-is-compatible

...
```

### What you expected to happen

For the package to download and the lint job to run successfully.

### How to reproduce it (as minimally and precisely as possible)

1. In this environment...
2. With this config...
3. Run '...'
4. See error...


### Environment

```shell
N/A
```

### Additional context

_No response_

--- a/packages/eslint-plugin-plugins/README.md
+++ b/packages/eslint-plugin-plugins/README.md
@@ -5,7 +5,7 @@
 ## How to install
 
 ```shell
-npm install eslint-plugin-plugins --save-dev
+npm install @grafana/eslint-plugin-plugins --save-dev
 ```
 
 ### Configure

--- a/packages/eslint-plugin-plugins/README.md
+++ b/packages/eslint-plugin-plugins/README.md
@@ -1,14 +1,39 @@
 # eslint-plugin-plugins
 
-is-compatible is a simple ESLint plugin that checks whether imports from any of the Grafana packages (`@grafana/ui`, `@grafana/data` and `@grafana/runtime`) from within a Grafana plugin source code exist in all the Grafana runtimes that the plugin is supposed to support (as specified in the `grafanaDependency` in `plugin.json`).
+`eslint-plugin-plugins` contains an ESLint rule that checks whether imports from any of the Grafana packages (`@grafana/ui`, `@grafana/data` and `@grafana/runtime`) from within a Grafana plugin source code exist in all the Grafana runtimes that the plugin is supposed to support (as specified in the `grafanaDependency` in `plugin.json`).
 
 ## How to install
 
 ```shell
-npm install @grafana/eslint-plugin-plugins --save-dev
+npm install eslint-plugin-plugins --save-dev
 ```
 
 ### Configure
+
+To determine the plugin's minimum supported Grafana version, the linter checks the `grafanaDependency` property in the plugin's `plugin.json`. By default, it looks for `plugin.json` in the `<projectRoot>/src` folder. If the file is located elsewhere or if you're using a monorepo with multiple plugins that have different `grafanaDependency` values, you can specify `minGrafanaVersion` directly in the ESLint configuration.
+
+#### Flat config
+
+```js
+const grafanaPlugins = require('@grafana/eslint-plugin-plugins');
+
+module.exports = [
+  // ...other configs
+  {
+    files: ['src/**/*.{ts,tsx}'],
+    plugins: { '@grafana/plugins': grafanaPlugins },
+    rules: {
+      '@grafana/plugins/import-is-compatible': [
+        'warn',
+        // optionally pass the minimum supported version
+        // { minGrafanaVersion: '10.3.0' },
+      ],
+    },
+  },
+];
+```
+
+#### Legacy config
 
 Add the following to your Grafana plugin's `.eslintrc`:
 
@@ -17,19 +42,11 @@ Add the following to your Grafana plugin's `.eslintrc`:
   ...
   "plugins": ["@grafana/eslint-plugin-plugins"],
   "rules": {
-    "@grafana/plugins/import-is-compatible": ["warn"]
-  }
-}
-```
-
-To determine the plugin's minimum supported Grafana version, the linter checks the `grafanaDependency` property in the plugin's `plugin.json`. By default, it looks for `plugin.json` in the `<projectRoot>/src` folder. If the file is located elsewhere or if you're using a monorepo with multiple plugins that have different `grafanaDependency` values, you can specify `minGrafanaVersion` directly in the ESLint configuration.
-
-```js
-{
-  ...
-  "plugins": ["@grafana/eslint-plugin-plugins"],
-  "rules": {
-    "@grafana/plugins/import-is-compatible": ["warn", { "minGrafanaVersion": "10.3.0" }]
+    "@grafana/plugins/import-is-compatible": [
+      "warn"
+      // optionally pass the minimum supported version
+      // { minGrafanaVersion: '10.3.0' },
+    ]
   }
 }
 ```

--- a/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
+++ b/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
@@ -2,7 +2,7 @@ import { join } from 'path';
 import { existsSync, mkdirSync } from 'fs';
 import { tmpdir } from 'os';
 import { execSync } from 'child_process';
-import { lt } from 'semver';
+import { lt, gt } from 'semver';
 import { getExportInfo } from './tscUtils';
 import { ExportInfo } from './types';
 
@@ -28,7 +28,7 @@ export function downloadPackages(tempDir: string, version: string) {
     );
   } else {
     packages.forEach((pkgName) => {
-      let typesFileUrl = `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/index.d.ts`;
+      let typesFileUrl = getPackageDownloadUrl(pkgName, version);
       let downloadPath = join(tempDir, 'node_modules', pkgName);
       mkdirSync(downloadPath, { recursive: true });
 
@@ -57,6 +57,17 @@ export function downloadPackages(tempDir: string, version: string) {
       }
     });
   }
+}
+
+function getPackageDownloadUrl(pkgName: string, version: string) {
+  if (gt(version, '12.1.0')) {
+    return `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/types/index.d.ts`;
+  }
+  if (gt(version, '11.0.0')) {
+    return `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/esm/index.d.mts`;
+  }
+
+  return `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/index.d.ts`;
 }
 
 function getPackageExportPaths(tempDir: string): Record<string, string> {

--- a/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
+++ b/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
@@ -74,6 +74,9 @@ function getPackageExportPaths(tempDir: string, minGrafanaVersion: string): Reco
         const packagePath = join(tempDir, 'node_modules', pkg);
         const packageJson = JSON.parse(readFileSync(join(packagePath, 'package.json'), 'utf8'));
         const typesPath = packageJson.types;
+        if (typeof typesPath !== 'string' || !typesPath) {
+          throw new Error(`Missing or invalid "types" field in ${pkg}'s package.json`);
+        }
         return {
           ...acc,
           [pkg]: join(packagePath, typesPath),

--- a/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
+++ b/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
@@ -1,5 +1,5 @@
 import { join } from 'path';
-import { existsSync, mkdirSync } from 'fs';
+import { existsSync, mkdirSync, readFileSync } from 'fs';
 import { tmpdir } from 'os';
 import { execSync } from 'child_process';
 import { lt, gte } from 'semver';
@@ -17,7 +17,7 @@ export function downloadPackages(tempDir: string, version: string) {
   console.log(`Please wait... downloading Grafana types information for version ${version}.`);
   mkdirSync(tempDir, { recursive: true });
 
-  if (lt(version, '9.2.0')) {
+  if (lt(version, '9.2.0') || gte(version, '12.1.0')) {
     execSync(
       `npm install ${packages.join(
         `@${version} `
@@ -60,17 +60,29 @@ export function downloadPackages(tempDir: string, version: string) {
 }
 
 function getPackageDownloadUrl(pkgName: string, version: string) {
-  if (gte(version, '12.1.0')) {
-    return `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/types/index.d.ts`;
-  }
   if (gte(version, '11.0.0')) {
-    return `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/esm/index.d.mts`;
+    return `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/cjs/index.d.cts`;
   }
 
   return `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/index.d.ts`;
 }
 
-function getPackageExportPaths(tempDir: string): Record<string, string> {
+function getPackageExportPaths(tempDir: string, minGrafanaVersion: string): Record<string, string> {
+  if (gte(minGrafanaVersion, '12.1.0')) {
+    try {
+      return packages.reduce<Record<string, string>>((acc, pkg) => {
+        const packagePath = require.resolve(pkg, { paths: [tempDir] });
+        const packageJson = JSON.parse(readFileSync(join(packagePath, 'package.json'), 'utf8'));
+        const typesPath = packageJson.types;
+        return {
+          ...acc,
+          [pkg]: join(packagePath, typesPath),
+        };
+      }, {});
+    } catch (error) {
+      throw new Error(`Failed to get package export paths: ${error}`);
+    }
+  }
   return packages.reduce(
     (acc, pkg) => ({
       ...acc,
@@ -87,7 +99,7 @@ export function getPackageExports(minGrafanaVersion: string): Record<string, Exp
     downloadPackages(tempDir, minGrafanaVersion);
   }
 
-  const packagePaths = getPackageExportPaths(tempDir);
+  const packagePaths = getPackageExportPaths(tempDir, minGrafanaVersion);
 
   return Object.entries(packagePaths).reduce(
     (acc, [pkg, path]) => ({

--- a/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
+++ b/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
@@ -2,7 +2,7 @@ import { join } from 'path';
 import { existsSync, mkdirSync } from 'fs';
 import { tmpdir } from 'os';
 import { execSync } from 'child_process';
-import { lt, gt } from 'semver';
+import { lt, gte } from 'semver';
 import { getExportInfo } from './tscUtils';
 import { ExportInfo } from './types';
 
@@ -60,10 +60,10 @@ export function downloadPackages(tempDir: string, version: string) {
 }
 
 function getPackageDownloadUrl(pkgName: string, version: string) {
-  if (gt(version, '12.1.0')) {
+  if (gte(version, '12.1.0')) {
     return `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/types/index.d.ts`;
   }
-  if (gt(version, '11.0.0')) {
+  if (gte(version, '11.0.0')) {
     return `https://cdn.jsdelivr.net/npm/${pkgName}@${version}/dist/esm/index.d.mts`;
   }
 

--- a/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
+++ b/packages/eslint-plugin-plugins/src/rules/getPackageExports.ts
@@ -71,7 +71,7 @@ function getPackageExportPaths(tempDir: string, minGrafanaVersion: string): Reco
   if (gte(minGrafanaVersion, '12.1.0')) {
     try {
       return packages.reduce<Record<string, string>>((acc, pkg) => {
-        const packagePath = require.resolve(pkg, { paths: [tempDir] });
+        const packagePath = join(tempDir, 'node_modules', pkg);
         const packageJson = JSON.parse(readFileSync(join(packagePath, 'package.json'), 'utf8'));
         const typesPath = packageJson.types;
         return {

